docs/design/tanka.md

```diff
@@ -0,0 +1,223 @@
+beyond application boundaries is already enabled by `jsonnet-bundler`.
+
+While it is possible to mimic environments with native `jsonnet`, it falls short
+when it comes to actually applying it to the correct cluster. The raw `json`
+being returned by the compiler needs reconciling and it must be made sure it is
+applied to the correct cluster.
+
+This leaves us effectively with **`jsonnet`** and **[Environments](#environments)**
+
+## Code Reuse
+By using [`jsonnet`](https://jsonnet.org) as the underlying data templating
+language, tanka dynamic reusing of code, just like real programming languages do:
+
+### Imports
+`jsonnet` is able to
+[import](https://jsonnet.org/learning/tutorial.html#imports) other jsonnet
+snippets into the current scope, which in turn allows to refactor commonly used code
+into shared libraries.
+
+### Bundle
+Once a library becomes general enough to be used beyond project or
+even domain boundaries, it is a common practice in programming languages to have
+shared dependencies.
+
+In the `jsonnet` world, this is provided by
+[`jsonnet-bundler`](https://github.com/jsonnet-bundler/jsonnet-bundler), which maintains a
+`vendor` folder (the bundle) with all required libraries ready to be imported,
+much like older versions of `go` used to (>1.11) do.  
+This procedure integrates smoothly with tanka, because `vendor` is on the [`JPATH`](#jpath).
+
+### `JPATH`
+To enable a predictable developer experience, tanka uses clear rules to define
+how importing works.
+
+Imports are relative to the `JPATH`. The earlier a directory appears in the
+`JPATH`, the higher it's precedence is.
+
+To set it up, tanka makes use of the following directoies:
+
+| Name             | Identifier         | Description                                                                                                                           |
+|------------------|--------------------|---------------------------------------------------------------------------------------------------------------------------------------|
+| `rootDir`        | `jsonnetfile.json` | Every file in the tree of this folder is considered part of the project. Much like `git` has the one directory with the `.git` folder |
+| `rootDir/vendor` |                    | Populated with shared dependencies by `jsonnet-bundler`                                                                               |
+| `baseDir/lib`    |                    | Code that is only re-used in-tree may be put here                                                                                     |
+| `baseDir`        | `main.jsonnet`     | Environment specific code can be put here. The special `main.jsonnet` is the entry point for the evaluation                           |
+
+To resolve the `JPATH`, tanka first traverses the directory tree *upwards*, to
+find a `jsonnetfile.json`, which marks the `rootDir`. Reaching `/` without a
+match will result in an error.
```

Comment:
Why do we need a `jsonnetfile.json`? Some setups won't have vendored libraries. Such a setup would then not work without an empty file.

Reply:
answering my own comment having read the whole file. I suspect this file defines where the `/lib` etc directories are to be found.